### ISSUE TITLE
re-factoring forecaster.update, update_predict, update_predict_single

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#982]
The rolling prediction interface of forecasters is currently confusing and unintuitive.

I suggest we re-factor this as follows:

* `update` remains as is currently
* `update_predict` becomes what `update_predict_single` is currently, a short-hand for "one `update` then one `predict`
* current functionality in `update_predict` gets replaced by an `update_predict_playback`, as below.

`update_predict_playback(y, cutoffs, X=None, fh=None, update_params=True) -> list`

`y` is the full series; `cutoffs` (a list of indices) are indices on the same axis as `y` (but not necessarily a sub-set of `y.index).
It is assumed (and asserted/tested) that `cutoff` is monotonous. `fh` must be relative.

The return is a list of same length as `cutoffs`; the 0-th element is the return of `predict` after `fit` to `y[:cutoffs[0]]` (temporal slice); the i-th element is the return of the i-th `update_predict`, with `y[cutoffs[i-1]:cutoffs[i]]`.
All predictions are with the `fh` passed.

**Comment by fkiraly** [comment]
@mloning, @aiwalter, thoughts?

**Comment by Lovkush-A** [comment]
maybe `..._multiple` instead of `..._playback`? You are making multiple updates and predicts, and it pairs well with the method name `update_predict_single`

**Comment by fkiraly** [comment]
>and it pairs well with the method name `update_predict_single`

but the proposal includes the removal of `update_predict_single`.
Further, it should be clear through naming that multiple application is basically the key step in backtesting, i.e., you already need your entire data ready. "multiple" isn't so clear here.

**Comment by mloning** [comment]
> * `update` remains as is currently
> * `update_predict` becomes what `update_predict_single` is currently, a short-hand for "one `update` then one `predict`

I think we have discussed this before, I fully agree. 

> * current functionality in `update_predict` gets replaced by an `update_predict_playback`, as below.

* I prefer @Lovkush-A's suggestion: `update_predict_multiple` or `update_predict_cv` would perhaps be clearer still. 
* What's the thinking behind changing the signature from `cv` to `cutoffs`? Don't CV objects encode different sets of cutoffs? Note that there is a `CutoffSplitter` which you can pass a list of cutoffs and it will do what you suggest. 
* More generally, my preference is to remove this method entirely and instead rely on stand-alone functions like `evaluate` for backtesting. To me, the implementation inside the class isn't very straightforward and most users are more familiar with functions similar to `cross_val_predict` from scikit-learn. 

Note that we need to run deprecation cycles when we remove this.

**Comment by Lovkush-A** [comment]
* I think the change from `cv` to `cutoffs` is because the name `cv` itself is not the best name. `cv` sounds like cross-validation, which sounds like it is some kind of function/method.

* I am also in favour of just removing the `update_predict_playback` method entirely, and instead having cross validation as a separate function as part of validation toolbox. However, this preference is biased by not knowing anything about current state of validation in sktime, or the pros and cons of having method within forecaster objects versus method outside forecaster objects.

**Comment by mloning** [comment]
@Lovkush-A note that `cv` is used in scikit-learn as an argument name for cross-validation iterators (see e.g. `GridSearchCV`) which is also what we pass here.

**Comment by Lovkush-A** [comment]
> @Lovkush-A note that `cv` is used in scikit-learn as an argument name for cross-validation iterators (see e.g. the constructor of `GridSearchCV`) which is also what we pass here.

Just checked the docs of sklearn. In my view, we should deviate from sklearn variable name choice in this instance, because I think it is significantly unclear.

**Comment by mloning** [comment]
@Lovkush-A I think most sktime users will be familiar with scikit-learn. What is unclear about, the suggestion in the name "cross-validation" that it's a function?  What would you call it?

**Comment by Lovkush-A** [comment]
@mloning 

> I think most sktime users will be familiar with scikit-learn.

Are most sktime users familiar enough with sklearn to know how `cv` is used in sklearn? I can only speak for myself and I did not know. I very well could be in the minority!

> What is unclear about, the suggestion in the name "cross-validation" that it's a function? What would you call it?

* To me, 'cv' sounds like it would be a function which carries out cross validation.
* Within GridSearchCV, I would prefer something like 'cv_strategy' or 'cv_type' or similar
* Within update_predict_playback (or whatever we will call it), my suggested alternative would depend on the possible values the `cv` variable can take. If the possible values are a list/series/index of times, then I prefer something like `cutoffs`.

**Comment by mloning** [comment]
I like `cv_strategy`, I agree it's clearer than just `cv`, but prefer not move away from scikit-learn.

**Comment by Lovkush-A** [comment]
@fkiraly @mloning. To summarise so far.

Things we agree on:
* agree on keeping `update`
* agree on renaming `update_predict_single` to `update_predict`, and have it correspond to `one update and then one predict`

Things we disagree on / that are not decided:
* Whether we keep a `update_predict_playback` method, or remove it as a forecaster method and replace it with something like `cross_val_predict`
* If we keep it, what the name of method should be.
* The name of `cv` variable

**Comment by fkiraly** [comment]
>What's the thinking behind changing the signature from cv to cutoffs? Don't CV objects encode different sets of cutoffs? Note that there is a `CutoffSplitter` which you can pass a list of cutoffs and it will do what you suggest.

I'm slightly confused about the intention of using `cv` objects that are not expanding windows. To me it seams this would dispatch to different `fit`/`predict` combinations, but what I think the method should be doing is repeated calls to `update` and `predict`. Cross-validation (temporal) other than continuous updating I would defer to something outside, like `evaluate`. Rolling predictions I would prefer to see inside the estimator.

**Comment by Lovkush-A** [comment]
To check, is this correct:

* temporal cross validation is multiple *fit* and predict cycles
* update_predict_playback is multiple *update* and predict cycles

If yes, for me, these two things are very similar (especially given that default update is just to re-fit) and hence should have a similar interface/naming conventions, etc.  They both appear to be different strategies for doing some kind of evaluation on past data.

**Comment by mloning** [comment]
We have summarised the different cross-validation schemes here: https://github.com/sktime/enhancement-proposals/blob/main/steps/01_forecasting_api/model-evaluation-tuning.md

**Comment by Lovkush-A** [comment]
> We have summarised the different cross-validation schemes

1. To check understanding, Franz's proposal for `update_predict_playback` matches up with "expanding window with an initial window" from the scheme?
2. If yes for 1., then it would be odd to have one of the five strategies implemented locally in the estimator, and the other four implemented externally.
3. Have the five cv schemes been implemented? It looks like there was not a decision made on the three design options, but I could have skimmed past it.

**Comment by fkiraly** [comment]
>To check understanding, Franz's proposal for update_predict_playback matches up with "expanding window with an initial window" from the scheme?

In the sense of what the method is allowed to use, yes. However, the method could be a moving window strategy, in which case the "true" scheme would correspond to moving window splitting.

>If yes for 1., then it would be odd to have one of the five strategies implemented locally in the estimator, and the other four implemented externally.

As far as I know, the expanding window one is the only one compatible with the `update`/`predict` sequence from the unified interface, since there is no `forget_the_past` method. Insofar it is the only one that mimicks continuous deployment usage.

Note that any CV method where training sets "lose" data points will have to rely on re-fitting, not updating.

**Comment by Lovkush-A** [comment]
Hmm. I am more confused now than before... 

> However, the method could be a moving window strategy

But in your specs at the top, what you suggested for `predict_update_playback` is more like expanding window than moving window, I think. Or have I misunderstood? For moving window, you need to specify length of window and how far window moves along each time. For expanding window, you need to specify cutoffs, which is what you have in the specs.

> As far as I know, the expanding window one is the only one compatible with the `update`/`predict` sequence from the unified interface, since there is no `forget_the_past` method. Insofar it is the only one that mimicks continuous deployment usage.

I am struggling to parse this paragraph.
* In the list of schemes (here: https://github.com/sktime/enhancement-proposals/blob/main/steps/01_forecasting_api/model-evaluation-tuning.md), there are two schemes that use update and predict, namely, schemes 3 and 5
* In your first paragraph, you say 'the method could be a moving window' and in this paragraph you are arguing that 'the expanding window is the only one'. This is confusing to me.
* I do not know how the existence of a `forget_the_past` method is relevant.
* Why do we want a forecaster to be able to mimick continuous deployment usage?
